### PR TITLE
Finished Authentication-collaborative project.

### PR DESCRIPTION
**Pull request by Fig77**
Worked on collaborative project number 6, of the RoR section [**Authentication**](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication).

This PR resolves the following points: 
* Basic Setup of Rails project.
* Setup for models Post and User
* Included security standards for password using bcrypt
* Implementation of SessionController
* Built pertinent forms for login and post.
* Add pertinent restrictions so nonusers could not see names of people who posted.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Rails.application.routes.draw do
+  root 'posts#index'
+  get 'posts', to: 'posts#new', as: 'posts'
+  post 'posts', to: 'posts#create'
+  get 'posts/index'
```
Comment (Torres-ssf) [inline config/routes.rb]: You can easily use posts resources here: `resources :posts, only: [:new, :create, :index]`

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  get 'login', to: 'sessions#new', as: 'users'
+  post 'login', to: 'sessions#create'
+  get 'show', to: 'sessions#show', as: 'show'
```
Comment (Torres-ssf) [inline config/routes.rb]: Why did you create a show session route? :thinking:

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  has_many :post, class_name: 'Post', foreign_key: 'idUser'
```
Comment (Torres-ssf) [inline app/models/user.rb]: Post must be in plural here
`has_many :posts...`

**Comment by Fig77** [comment]
Hi Sergio, 
Thanks for your kind review. We fixed all the issues regarding our code. Let us know if there is anything else we can do.

The reason that we did not change the routes, it's because we wanted to have a different path instead of localhost:3000/users/login, we want to have just localhost:3000/login.

File: app/views/sessions/_loginForm.html.erb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<%= form_for @user do |f| %>
+  <%= f.text_field :email, :placeholder => 'email' %>
+  <%= f.text_field :password, :placeholder => 'password' %>
```
Comment (maelfosso) [inline app/views/sessions/_loginForm.html.erb]: **password_field** instead of `text_field`

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
+class ApplicationController < ActionController::Base
+  protect_from_forgery with: :exception
+
+  def permanent_coockie(user)
+    temp = SecureRandom.urlsafe_base64
+    temp = Digest::SHA1.hexdigest temp
+    user.update(token: temp)
+    cookies.permanent[:remember_token] = temp
+  end
```
Comment (maelfosso) [inline app/controllers/application_controller.rb]: You also have to set the current user.

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
+class ApplicationController < ActionController::Base
+  protect_from_forgery with: :exception
+
+  def permanent_coockie(user)
+    temp = SecureRandom.urlsafe_base64
+    temp = Digest::SHA1.hexdigest temp
```
Comment (maelfosso) [inline app/controllers/application_controller.rb]: That snippet of code appears twice. I suggest you write a method in User model
```ruby
def self.generate_token
    temp = SecureRandom.urlsafe_base64
    temp = Digest::SHA1.hexdigest temp
end
```

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    cookies.permanent[:remember_token] = temp
+  end
+
+  def retriver
+    @current_user ||= User.find_by_token(cookies.permanent[:remember_token]) unless sign_in?
+  end
+
+  def current_user
+    @current_user ||= User.find_by(id: session[:user_id]) if session[:user_id]
+  end
```
Comment (maelfosso) [inline app/controllers/application_controller.rb]: No **session**!!!

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+  def permanent_coockie(user)
+    temp = SecureRandom.urlsafe_base64
+    temp = Digest::SHA1.hexdigest temp
+    user.update(token: temp)
+    cookies.permanent[:remember_token] = temp
+  end
+
+  def retriver
+    @current_user ||= User.find_by_token(cookies.permanent[:remember_token]) unless sign_in?
```
Comment (maelfosso) [inline app/controllers/application_controller.rb]: No need of `unless sign_in?`

File: app/controllers/sessions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+  def session_paramns
+    params.require(:user).permit(:token)
+  end
+
+  def sign_in(user)
+    session[:user_id] = user.id
+    session[:token] = user.token
+  end
```
Comment (maelfosso) [inline app/controllers/sessions_controller.rb]: not needed !!!
No use of `session`

File: app/controllers/sessions_controller.rb [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+  def sign_out
+    session.delete(:user_id)
+    session.delete(:token)
+    cookies.permanent[:remember_token] = ''
```
Comment (maelfosso) [inline app/controllers/sessions_controller.rb]: `cookies.delete(:remember_token)` instead

File: app/controllers/sessions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+  def sign_in(user)
+    session[:user_id] = user.id
+    session[:token] = user.token
+  end
+
+  def sign_out
+    session.delete(:user_id)
+    session.delete(:token)
```
Comment (maelfosso) [inline app/controllers/sessions_controller.rb]: No use of `session`

File: app/controllers/sessions_controller.rb (hunk 10 above)
Comment (maelfosso) [reply app/controllers/sessions_controller.rb]: You also have to set the current_user to nil

**Comment by Dandush03** [comment]
Hey Mael :hand:,

we fixed most of the issues explained above :+1:  :muscle:, but we encounter 2 suggestion that isn't in the project specification; nor in microverse, nor in the Odin project :disappointed:. 

  * There is no route under '/posts/new' as the [error display above](https://github.com/Fig77/Authentication-collaborative/pull/1#pullrequestreview-375641224) instead is you can access it by '/posts'
  * There is now project specification calling us to not use session. please advice, [check the image below](#Project), and before this project microvers path even ask us to make a whole reading regarding this matter [check this link](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/37867493/)

---

#Project

---
Best Regards, 

:facepunch: The Gozzillas Team :bowtie:

File: app/controllers/sessions_controller.rb (hunk 11 above)
Comment (maelfosso) [reply app/controllers/sessions_controller.rb]: No use of session

**Comment by Dandush03** [comment]
Hey Again!

We Just finished deleting all the sessions.

## Issue with deleting cookies from User.

if we do so, we will encounter a big issue. since everything is based on cookies the app will search for a cookie that matches the permanent_cockie therefore if I had a user with cookie = nil, and permanent_cockie is also = to nil then it will automatically get logged in. Any advice?

Please let us know if there is anything else we can do to get this project to approve

best regards

:facepunch: The Gozzillas Team :bowtie:

**Pull request merged**